File: services/ai-evaluation-service/src/handlers/evaluations.rs
use std::{collections::BTreeSet, time::Instant};

use axum::{
    Json,
    extract::State,
};
use chrono::Utc;
use serde_json::{Value, json};
use sqlx::{query_as, types::Json as SqlJson};
use uuid::Uuid;

use crate::{
    AppState,
    domain::{evaluation, llm},
    models::{
        conversation::{
            ChatAttachment, EvaluateGuardrailsRequest, EvaluateGuardrailsResponse,
            GuardrailVerdict, LlmUsageSummary, ProviderBenchmarkRequest,
            ProviderBenchmarkResponse, ProviderBenchmarkResult, ProviderBenchmarkScore,
        },
        provider::{LlmProvider, ProviderRow},
    },
};

use super::{ServiceResult, bad_request, db_error, not_found};

async fn load_provider_rows(db: &sqlx::PgPool) -> Result<Vec<ProviderRow>, sqlx::Error> {
    query_as::<_, ProviderRow>(
        r#"
        SELECT
            id,
            name,
            provider_type,
            model_name,
            endpoint_url,
            api_mode,
            credential_reference,
            enabled,
            load_balance_weight,
            max_output_tokens,
            cost_tier,
            tags,
            route_rules,
            health_state,
            created_at,
            updated_at
        FROM ai_providers
        ORDER BY updated_at DESC, created_at DESC
        "#,
    )
    .fetch_all(db)
    .await
}

fn summarize_title(content: &str) -> String {
    let mut chars = content.trim().chars();
    let title = chars.by_ref().take(60).collect::<String>();
    if chars.next().is_some() {
        format!("{title}...")
    } else if title.is_empty() {
        "New benchmark".to_string()
    } else {
        title
    }
}

fn preview_text(content: &str, limit: usize) -> String {
    let mut chars = content.trim().chars();
    let preview = chars.by_ref().take(limit).collect::<String>();
    if chars.next().is_some() {
        format!("{preview}...")
    } else {
        preview
    }
}

fn attachment_context(attachments: &[ChatAttachment]) -> String {
    if attachments.is_empty() {
        return "none".to_string();
    }

    attachments
        .iter()
        .map(|attachment| {
            let label = attachment
                .name
                .as_deref()
                .filter(|value| !value.trim().is_empty())
                .unwrap_or("attachment");
            match attachment.kind.as_str() {
                "image_url" => format!(
                    "- {label}: image url {}",
                    attachment.url.as_deref().unwrap_or("missing-url")
                ),
                "image_base64" => format!(
                    "- {label}: embedded {} image",
                    attachment.mime_type.as_deref().unwrap_or("unknown")
                ),
                _ => format!(
                    "- {label}: {}",
                    attachment.text.as_deref().unwrap_or("text attachment")
                ),
            }
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn required_modalities(attachments: &[ChatAttachment]) -> Vec<String> {
    let mut modalities = vec!["text".to_string()];
    if attachments
        .iter()
        .any(|attachment| attachment.kind.starts_with("image"))
    {
        modalities.push("image".to_string());
    }
    modalities
}

fn modality_label(required_modalities: &[String]) -> &'static str {
    if required_modalities
        .iter()
        .any(|modality| modality.eq_ignore_ascii_case("image"))
    {
        "image+text"
    } else {
        "text"
    }
}

fn privacy_reason(guardrail: &GuardrailVerdict, require_private_network: bool) -> Option<String> {
    if require_private_network {
        Some("private network explicitly requested".to_string())
    } else if guardrail
        .flags
        .iter()
        .any(|flag| flag.kind.starts_with("pii_"))
    {
        Some("PII detected in prompt, preferring private-network providers".to_string())
    } else {
        None
    }
}

fn usage_summary(
    provider: &LlmProvider,
    prompt_tokens: i32,
    completion_tokens: i32,
    latency_ms: i32,
    cache_hit: bool,
) -> LlmUsageSummary {
    let total_tokens = prompt_tokens.max(0) + completion_tokens.max(0);

    LlmUsageSummary {
        prompt_tokens,
        completion_tokens,
        total_tokens,
        estimated_cost_usd: evaluation::estimated_cost_usd(
            provider,
            prompt_tokens,
            completion_tokens,
            cache_hit,
        ),
        latency_ms,
        network_scope: provider.route_rules.network_scope.clone(),
        cache_hit,
    }
}

async fn record_usage_event(
    db: &sqlx::PgPool,
    provider_id: Uuid,
    request_kind: &str,
    use_case: &str,
    modality: &str,
    usage: &LlmUsageSummary,
    benchmark_group_id: Option<Uuid>,
    metadata: Value,
) -> Result<(), sqlx::Error> {
    sqlx::query(
        r#"
        INSERT INTO ai_llm_usage_events (
            id,
            provider_id,
            conversation_id,
            request_kind,
            use_case,
            network_scope,
            modality,
            cache_hit,
            prompt_tokens,
            completion_tokens,
            total_tokens,
            estimated_cost_usd,
            latency_ms,
            benchmark_group_id,
            metadata
        )
        VALUES ($1, $2, NULL, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
        "#,
    )
    .bind(Uuid::now_v7())
    .bind(provider_id)
    .bind(request_kind)
    .bind(use_case)
    .bind(&usage.network_scope)
    .bind(modality)
    .bind(usage.cache_hit)
    .bind(usage.prompt_tokens)
    .bind(usage.completion_tokens)
    .bind(usage.total_tokens)
    .bind(usage.estimated_cost_usd as f64)
    .bind(usage.latency_ms)
    .bind(benchmark_group_id)
    .bind(SqlJson(metadata))
    .execute(db)
    .await?;

    Ok(())
}

pub async fn benchmark_providers(
    State(state): State<AppState>,
    Json(body): Json<ProviderBenchmarkRequest>,
) -> ServiceResult<ProviderBenchmarkResponse> {
    if body.prompt.trim().is_empty() {
        return Err(bad_request("benchmark prompt is required"));
    }

    let prompt_guardrail = llm::guardrails::evaluate_text(&body.prompt);
    if prompt_guardrail.blocked {
        return Err(bad_request(
            "benchmark prompt is blocked by guardrails; sanitize it before benchmarking",
        ));
    }

    let providers = load_provider_rows(&state.db)
        .await
        .map_err(|cause| db_error(&cause))?
        .into_iter()
        .map(Into::into)
        .collect::<Vec<LlmProvider>>();
    if providers.is_empty() {
        return Err(not_found("no AI providers configured"));
    }

    let candidate_providers = if body.provider_ids.is_empty() {
        providers.clone()
    } else {
        providers
            .iter()
            .filter(|provider| body.provider_ids.contains(&provider.id))
            .cloned()
            .collect::<Vec<_>>()
    };
    if candidate_providers.is_empty() {
        return Err(not_found(
            "no benchmark providers matched the requested ids",
        ));
    }

    let required_modalities = required_modalities(&body.attachments);
    let privacy_reason = privacy_reason(&prompt_guardrail, body.require_private_network);
    let routed = llm::gateway::route_providers(
        &candidate_providers,
        None,
        &body.use_case,
        &required_modalities,
        body.require_private_network,
        privacy_reason.is_some(),
    );
    if body.require_private_network && routed.is_empty() {
        return Err(bad_request(
            "no private-network AI provider is configured for this benchmark",
        ));
    }
    if routed.is_empty() {
        return Err(not_found("no eligible providers support this benchmark"));
    }

    let benchmark_group_id = Uuid::now_v7();
    let system_prompt = body.system_prompt.unwrap_or_else(|| {
        "You are an enterprise AI benchmark harness. Answer the user prompt clearly and concretely."
            .to_string()
    });
    let prompt_used = format!(
        "{}\n\nUser request: {}\n\nAttachments:\n{}",
        system_prompt,
        prompt_guardrail.redacted_text,
        attachment_context(&body.attachments)
    );
    let mut results = Vec::new();

    for provider in routed {
        let started_at = Instant::now();
        let completion = llm::runtime::complete_text(
            &state.http_client,
            &provider,
            &system_prompt,
            &body.prompt,
            &body.attachments,
            0.2,
            body.max_tokens,
        )
        .await;
        let latency_ms = started_at.elapsed().as_millis().min(i32::MAX as u128) as i32;

        match completion {
            Ok(completion) => {
                let prompt_tokens = completion
                    .prompt_tokens
                    .max(llm::gateway::estimate_tokens(&prompt_used));
                let completion_tokens = completion
                    .completion_tokens
                    .max(llm::gateway::estimate_tokens(&completion.text));
                let mut usage = usage_summary(
                    &provider,
                    prompt_tokens,
                    completion_tokens,
                    latency_ms,
                    false,
                );
                usage.total_tokens = completion
                    .total_tokens
                    .max(prompt_tokens + completion_tokens);
                let reply_guardrail = llm::guardrails::evaluate_text(&completion.text);
                record_usage_event(
                    &state.db,
                    provider.id,
                    "benchmark",
                    &body.use_case,
                    modality_label(&required_modalities),
                    &usage,
                    Some(benchmark_group_id),
                    json!({
                        "rubric_keywords": body.rubric_keywords,
                        "provider_name": provider.name,
                    }),
                )
                .await
                .map_err(|cause| db_error(&cause))?;

                results.push(ProviderBenchmarkResult {
                    provider_id: provider.id,
                    provider_name: provider.name,
                    network_scope: usage.network_scope.clone(),
                    reply_preview: preview_text(&completion.text, 280),
                    prompt_tokens: usage.prompt_tokens,
                    completion_tokens: usage.completion_tokens,
                    total_tokens: usage.total_tokens,
                    estimated_cost_usd: usage.estimated_cost_usd,
                    latency_ms: usage.latency_ms,
                    cache_hit: false,
                    guardrail: reply_guardrail,
                    score: ProviderBenchmarkScore {
                        quality: 0.0,
                        latency: 0.0,
                        cost: 0.0,
                        safety: 0.0,
                        overall: 0.0,
                    },
                    error: None,
                });
            }
            Err(error) => {
                results.push(ProviderBenchmarkResult {
                    provider_id: provider.id,
                    provider_name: provider.name,
                    network_scope: provider.route_rules.network_scope.clone(),
                    reply_preview: String::new(),
                    prompt_tokens: 0,
                    completion_tokens: 0,
                    total_tokens: 0,
                    estimated_cost_usd: 0.0,
                    latency_ms,
                    cache_hit: false,
                    guardrail: GuardrailVerdict::default(),
                    score: ProviderBenchmarkScore {
                        quality: 0.0,
                        latency: 0.0,
                        cost: 0.0,
                        safety: 0.0,
                        overall: 0.0,
                    },
                    error: Some(error),
                });
            }
        }
    }

    let successful = results
        .iter()
        .filter(|result| result.error.is_none())
        .collect::<Vec<_>>();
    let min_latency = successful
        .iter()
        .map(|result| result.latency_ms as f32)
        .fold(f32::INFINITY, f32::min);
    let max_latency = successful
        .iter()
        .map(|result| result.latency_ms as f32)
        .fold(0.0, f32::max);
    let min_cost = successful
        .iter()
        .map(|result| result.estimated_cost_usd)
        .fold(f32::INFINITY, f32::min);
    let max_cost = successful
        .iter()
        .map(|result| result.estimated_cost_usd)
        .fold(0.0, f32::max);

    for result in &mut results {
        if result.error.is_some() {
            continue;
        }

        let quality = evaluation::quality_score(&result.reply_preview, &body.rubric_keywords);
        let safety = evaluation::safety_score(&result.guardrail);
        let latency =
            evaluation::normalized_score(result.latency_ms as f32, min_latency, max_latency, true);
        let cost =
            evaluation::normalized_score(result.estimated_cost_usd, min_cost, max_cost, true);
        result.score = ProviderBenchmarkScore {
            quality,
            latency,
            cost,
            safety,
            overall: evaluation::overall_benchmark_score(quality, safety, latency, cost),
        };
    }

    results.sort_by(|left, right| right.score.overall.total_cmp(&left.score.overall));
    let recommended_provider_id = results
        .iter()
        .find(|result| result.error.is_none())
        .map(|result| result.provider_id);

    Ok(Json(ProviderBenchmarkResponse {
        benchmark_group_id,
        use_case: body.use_case,
        prompt_excerpt: summarize_title(&body.prompt),
        required_modalities,
        requested_private_network: body.require_private_network,
        recommended_provider_id,
        results,
        created_at: Utc::now(),
    }))
}

pub async fn evaluate_guardrails(
    State(_state): State<AppState>,
    Json(body): Json<EvaluateGuardrailsRequest>,
) -> ServiceResult<EvaluateGuardrailsResponse> {
    if body.content.trim().is_empty() {
        return Err(bad_request("guardrail evaluation requires content"));
    }

    let verdict = llm::guardrails::evaluate_text(&body.content);
    let mut recommendations = BTreeSet::new();

    if verdict.blocked {
        recommendations
            .insert("Remove prompt-injection or toxic content before retrying.".to_string());
    }
    for flag in &verdict.flags {
        if flag.kind.starts_with("pii_") {
            recommendations
                .insert("Redact PII before routing prompts to external LLM providers.".to_string());
        }
    }
    if recommendations.is_empty() {
        recommendations
            .insert("No blocking issues detected; response is safe to continue.".to_string());
    }

    Ok(Json(EvaluateGuardrailsResponse {
        verdict: verdict.clone(),
        risk_score: evaluation::risk_score(&verdict),
        recommendations: recommendations.into_iter().collect(),
    }))
}

File: services/ai-evaluation-service/internal/handlers/evaluations.go
// Package handlers exposes the HTTP surface of ai-evaluation-service:
// /api/v1/evaluations/benchmark + /api/v1/guardrails/evaluate. Mirrors
// services/ai-evaluation-service/src/handlers/evaluations.rs verbatim.
//
// The benchmark path validates the prompt (guardrail-blocked prompts
// 400 before any provider runs), filters providers by the optional
// provider_ids list, routes the candidates via the gateway with
// modality + privacy filters, calls llm.CompleteText for each routed
// provider, records ai_llm_usage_events per success, scores
// quality/safety/latency/cost/overall, sorts overall desc, and picks
// the head as recommended_provider_id.
//
// The guardrails path is pure logic: runs llm.EvaluateText, derives
// the recommendations list (sorted, mirroring the Rust BTreeSet), and
// returns the verdict + risk score.
package handlers

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"sort"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/openfoundry/openfoundry-go/libs/ai-kernel-go/domain/evaluation"
	"github.com/openfoundry/openfoundry-go/libs/ai-kernel-go/domain/llm"
	"github.com/openfoundry/openfoundry-go/libs/ai-kernel-go/models"
)

// Handlers wires the evaluation HTTP surface. Pool is the pgx pool
// against the ai-evaluation-service database (which owns the
// ai_providers + ai_llm_usage_events tables). Runtime is the LLM
// dispatch — production wiring uses HTTPRuntime, tests inject
// FakeRuntime.
type Handlers struct {
	Pool    *pgxpool.Pool
	Runtime llm.Runtime
}

// completionRuntime returns the injected runtime, falling back to the
// real HTTP runtime when none is configured.
func (h *Handlers) completionRuntime() llm.Runtime {
	if h.Runtime != nil {
		return h.Runtime
	}
	return llm.HTTPRuntime{}
}

// errorResponse mirrors Rust handlers::mod::ErrorResponse — the
// canonical {"error": "..."} envelope.
type errorResponse struct {
	Error string `json:"error"`
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func writeError(w http.ResponseWriter, status int, msg string) {
	writeJSON(w, status, errorResponse{Error: msg})
}

func dbError(w http.ResponseWriter, err error) {
	if errors.Is(err, pgx.ErrNoRows) {
		writeError(w, http.StatusNotFound, "not found")
		return
	}
	writeError(w, http.StatusInternalServerError, "database operation failed")
}

// providerColumns matches libs/ai-kernel-go/handlers — same column
// order so scanProvider stays in sync with the kernel scanner.
const providerColumns = `id, name, provider_type, model_name, endpoint_url,
                         api_mode, credential_reference, enabled,
                         load_balance_weight, max_output_tokens,
                         cost_tier, tags, route_rules, health_state,
                         created_at, updated_at`

// rowScanner is the minimal pgx scanner interface accepted by
// scanProvider — both pgx.Row and pgx.Rows satisfy it.
type rowScanner interface {
	Scan(dest ...any) error
}

// scanProvider mirrors libs/ai-kernel-go/handlers.scanProvider — same
// JSON-decoded route_rules + health_state shape, same defaults.
func scanProvider(s rowScanner) (models.LlmProvider, error) {
	var p models.LlmProvider
	var credRef *string
	var tagsRaw, routeRulesRaw, healthRaw []byte
	if err := s.Scan(
		&p.ID, &p.Name, &p.ProviderType, &p.ModelName, &p.EndpointURL,
		&p.APIMode, &credRef, &p.Enabled,
		&p.LoadBalanceWeight, &p.MaxOutputTokens,
		&p.CostTier, &tagsRaw, &routeRulesRaw, &healthRaw,
		&p.CreatedAt, &p.UpdatedAt,
	); err != nil {
		return p, err
	}
	p.CredentialReference = credRef
	p.CredentialConfigured = credRef != nil && strings.TrimSpace(*credRef) != ""
	if len(tagsRaw) > 0 {
		_ = json.Unmarshal(tagsRaw, &p.Tags)
	}
	if p.Tags == nil {
		p.Tags = []string{}
	}
	if len(routeRulesRaw) > 0 {
		_ = json.Unmarshal(routeRulesRaw, &p.RouteRules)
	}
	if len(healthRaw) > 0 {
		_ = json.Unmarshal(healthRaw, &p.HealthState)
	}
	return p, nil
}

// loadProviderRows mirrors fn load_provider_rows — every configured
// provider ordered by updated_at desc, created_at desc.
func loadProviderRows(ctx context.Context, pool *pgxpool.Pool) ([]models.LlmProvider, error) {
	rows, err := pool.Query(ctx,
		`SELECT `+providerColumns+` FROM ai_providers
          ORDER BY updated_at DESC, created_at DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]models.LlmProvider, 0)
	for rows.Next() {
		p, err := scanProvider(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, p)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	return out, nil
}

// summarizeTitle mirrors fn summarize_title — first 60 runes of the
// trimmed content, "..." suffix if longer, "New benchmark" when empty.
func summarizeTitle(content string) string {
	trimmed := strings.TrimSpace(content)
	runes := []rune(trimmed)
	if len(runes) > 60 {
		return string(runes[:60]) + "..."
	}
	if len(runes) == 0 {
		return "New benchmark"
	}
	return string(runes)
}

// previewText mirrors fn preview_text — first `limit` runes of the
// trimmed content; appends "..." if truncated.
func previewText(content string, limit int) string {
	trimmed := strings.TrimSpace(content)
	runes := []rune(trimmed)
	if len(runes) > limit {
		return string(runes[:limit]) + "..."
	}
	return string(runes)
}

// attachmentContext mirrors fn attachment_context — formats
// attachments as "- <label>: …" lines for inclusion in the prompt
// echo. Returns "none" when the slice is empty.
func attachmentContext(attachments []models.ChatAttachment) string {
	if len(attachments) == 0 {
		return "none"
	}
	lines := make([]string, 0, len(attachments))
	for _, a := range attachments {
		label := "attachment"
		if a.Name != nil && strings.TrimSpace(*a.Name) != "" {
			label = *a.Name
		}
		switch a.Kind {
		case "image_url":
			url := "missing-url"
			if a.URL != nil {
				url = *a.URL
			}
			lines = append(lines, fmt.Sprintf("- %s: image url %s", label, url))
		case "image_base64":
			mime := "unknown"
			if a.MimeType != nil {
				mime = *a.MimeType
			}
			lines = append(lines, fmt.Sprintf("- %s: embedded %s image", label, mime))
		default:
			text := "text attachment"
			if a.Text != nil {
				text = *a.Text
			}
			lines = append(lines, fmt.Sprintf("- %s: %s", label, text))
		}
	}
	return strings.Join(lines, "\n")
}

// requiredModalities mirrors fn required_modalities — always includes
// "text"; appends "image" when any attachment kind starts with "image".
func requiredModalities(attachments []models.ChatAttachment) []string {
	out := []string{"text"}
	for _, a := range attachments {
		if strings.HasPrefix(a.Kind, "image") {
			out = append(out, "image")
			break
		}
	}
	return out
}

// modalityLabel mirrors fn modality_label.
func modalityLabel(required []string) string {
	for _, m := range required {
		if strings.EqualFold(m, "image") {
			return "image+text"
		}
	}
	return "text"
}

// privacyReason mirrors fn privacy_reason — "private network
// explicitly requested" when the body flag is set; the PII fallback
// when guardrails flagged a pii_* kind; nil otherwise.
func privacyReason(verdict models.GuardrailVerdict, requirePrivateNetwork bool) *string {
	if requirePrivateNetwork {
		s := "private network explicitly requested"
		return &s
	}
	for _, f := range verdict.Flags {
		if strings.HasPrefix(f.Kind, "pii_") {
			s := "PII detected in prompt, preferring private-network providers"
			return &s
		}
	}
	return nil
}

// usageSummary mirrors fn usage_summary — clamps negative tokens to
// zero before summing, computes EstimatedCostUSD via the kernel
// evaluation helper.
func usageSummary(provider models.LlmProvider, promptTokens, completionTokens, latencyMs int32, cacheHit bool) models.LlmUsageSummary {
	pt := promptTokens
	if pt < 0 {
		pt = 0
	}
	ct := completionTokens
	if ct < 0 {
		ct = 0
	}
	return models.LlmUsageSummary{
		PromptTokens:     promptTokens,
		CompletionTokens: completionTokens,
		TotalTokens:      pt + ct,
		EstimatedCostUSD: evaluation.EstimatedCostUSD(&provider, promptTokens, completionTokens, cacheHit),
		LatencyMs:        latencyMs,
		NetworkScope:     provider.RouteRules.NetworkScope,
		CacheHit:         cacheHit,
	}
}

// recordUsageEvent mirrors fn record_usage_event — best-effort insert
// into ai_llm_usage_events. The benchmark loop swallows the error so
// a metric-write hiccup doesn't kill the whole batch.
func recordUsageEvent(
	ctx context.Context,
	pool *pgxpool.Pool,
	providerID uuid.UUID,
	requestKind, useCase, modality string,
	usage models.LlmUsageSummary,
	benchmarkGroupID *uuid.UUID,
	metadata any,
) error {
	id, err := uuid.NewV7()
	if err != nil {
		id = uuid.New()
	}
	metadataJSON, _ := json.Marshal(metadata)
	_, err = pool.Exec(ctx,
		`INSERT INTO ai_llm_usage_events (
            id, provider_id, conversation_id, request_kind, use_case,
            network_scope, modality, cache_hit, prompt_tokens,
            completion_tokens, total_tokens, estimated_cost_usd,
            latency_ms, benchmark_group_id, metadata
         ) VALUES ($1, $2, NULL, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)`,
		id, providerID, requestKind, useCase,
		usage.NetworkScope, modality, usage.CacheHit, usage.PromptTokens,
		usage.CompletionTokens, usage.TotalTokens, usage.EstimatedCostUSD,
		usage.LatencyMs, benchmarkGroupID, metadataJSON,
	)
	return err
}

// BenchmarkProviders handles `POST /api/v1/evaluations/benchmark`.
// Mirrors fn benchmark_providers verbatim:
//
//  1. validate prompt + guardrail (block sanitises 400 if blocked)
//  2. load providers, optionally filtered by body.provider_ids
//  3. route via gateway with privacy + modality filters
//  4. call llm.CompleteText for each routed provider — capture
//     latency, tokens, error
//  5. record ai_llm_usage_events per success (best-effort)
//  6. score quality/safety/latency/cost/overall, sort overall desc,
//     pick the head as recommended_provider_id.
func (h *Handlers) BenchmarkProviders(w http.ResponseWriter, r *http.Request) {
	if h.Pool == nil {
		writeError(w, http.StatusServiceUnavailable, "ai-evaluation-service is not configured with a database")
		return
	}

	var body models.ProviderBenchmarkRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if strings.TrimSpace(body.Prompt) == "" {
		writeError(w, http.StatusBadRequest, "benchmark prompt is required")
		return
	}
	if strings.TrimSpace(body.UseCase) == "" {
		body.UseCase = models.DefaultBenchmarkUseCase
	}
	if body.MaxTokens <= 0 {
		body.MaxTokens = models.DefaultMaxTokens
	}

	promptVerdict := llm.EvaluateText(body.Prompt)
	if promptVerdict.Blocked {
		writeError(w, http.StatusBadRequest,
			"benchmark prompt is blocked by guardrails; sanitize it before benchmarking")
		return
	}

	ctx := r.Context()
	providers, err := loadProviderRows(ctx, h.Pool)
	if err != nil {
		dbError(w, err)
		return
	}
	if len(providers) == 0 {
		writeError(w, http.StatusNotFound, "no AI providers configured")
		return
	}

	candidates := providers
	if len(body.ProviderIDs) > 0 {
		filterSet := map[uuid.UUID]struct{}{}
		for _, id := range body.ProviderIDs {
			filterSet[id] = struct{}{}
		}
		filtered := make([]models.LlmProvider, 0, len(filterSet))
		for _, p := range providers {
			if _, ok := filterSet[p.ID]; ok {
				filtered = append(filtered, p)
			}
		}
		candidates = filtered
	}
	if len(candidates) == 0 {
		writeError(w, http.StatusNotFound, "no benchmark providers matched the requested ids")
		return
	}

	required := requiredModalities(body.Attachments)
	privacy := privacyReason(promptVerdict, body.RequirePrivateNetwork)
	routed := llm.RouteProviders(candidates, nil, body.UseCase, required,
		body.RequirePrivateNetwork, privacy != nil)
	if body.RequirePrivateNetwork && len(routed) == 0 {
		writeError(w, http.StatusBadRequest,
			"no private-network AI provider is configured for this benchmark")
		return
	}
	if len(routed) == 0 {
		writeError(w, http.StatusNotFound, "no eligible providers support this benchmark")
		return
	}

	benchmarkGroupID, err := uuid.NewV7()
	if err != nil {
		benchmarkGroupID = uuid.New()
	}

	systemPrompt := "You are an enterprise AI benchmark harness. Answer the user prompt clearly and concretely."
	if body.SystemPrompt != nil && strings.TrimSpace(*body.SystemPrompt) != "" {
		systemPrompt = *body.SystemPrompt
	}
	promptUsed := fmt.Sprintf("%s\n\nUser request: %s\n\nAttachments:\n%s",
		systemPrompt, promptVerdict.RedactedText, attachmentContext(body.Attachments))

	results := make([]models.ProviderBenchmarkResult, 0, len(routed))
	for _, provider := range routed {
		startedAt := time.Now()
		completion, completionErr := h.completionRuntime().CompleteText(ctx, llm.CompletionRequest{
			Provider:     &provider,
			SystemPrompt: systemPrompt,
			UserPrompt:   body.Prompt,
			Attachments:  body.Attachments,
			Temperature:  models.DefaultTemperature,
			MaxTokens:    body.MaxTokens,
		})
		latencyMs := int32(time.Since(startedAt).Milliseconds())
		if latencyMs < 0 {
			latencyMs = 0
		}

		if completionErr != nil {
			errStr := completionErr.Error()
			results = append(results, models.ProviderBenchmarkResult{
				ProviderID:       provider.ID,
				ProviderName:     provider.Name,
				NetworkScope:     provider.RouteRules.NetworkScope,
				ReplyPreview:     "",
				PromptTokens:     0,
				CompletionTokens: 0,
				TotalTokens:      0,
				EstimatedCostUSD: 0,
				LatencyMs:        latencyMs,
				CacheHit:         false,
				Guardrail:        models.DefaultGuardrailVerdict(),
				Score:            models.ProviderBenchmarkScore{},
				Error:            &errStr,
			})
			continue
		}

		promptTokens := completion.PromptTokens
		if est := llm.EstimateTokens(promptUsed); est > promptTokens {
			promptTokens = est
		}
		completionTokens := completion.CompletionTokens
		if est := llm.EstimateTokens(completion.Text); est > completionTokens {
			completionTokens = est
		}
		usage := usageSummary(provider, promptTokens, completionTokens, latencyMs, false)
		if completion.TotalTokens > usage.TotalTokens {
			usage.TotalTokens = completion.TotalTokens
		}

		replyVerdict := llm.EvaluateText(completion.Text)

		// Best-effort metric emission; the benchmark response stays
		// consistent even when the insert is rejected.
		_ = recordUsageEvent(ctx, h.Pool, provider.ID, "benchmark",
			body.UseCase, modalityLabel(required), usage, &benchmarkGroupID,
			map[string]any{
				"rubric_keywords": body.RubricKeywords,
				"provider_name":   provider.Name,
			})

		results = append(results, models.ProviderBenchmarkResult{
			ProviderID:       provider.ID,
			ProviderName:     provider.Name,
			NetworkScope:     usage.NetworkScope,
			ReplyPreview:     previewText(completion.Text, 280),
			PromptTokens:     usage.PromptTokens,
			CompletionTokens: usage.CompletionTokens,
			TotalTokens:      usage.TotalTokens,
			EstimatedCostUSD: usage.EstimatedCostUSD,
			LatencyMs:        usage.LatencyMs,
			CacheHit:         false,
			Guardrail:        replyVerdict,
			Score:            models.ProviderBenchmarkScore{},
		})
	}

	successful := make([]int, 0, len(results))
	for i, r := range results {
		if r.Error == nil {
			successful = append(successful, i)
		}
	}

	minLatency, maxLatency := float32(0), float32(0)
	minCost, maxCost := float32(0), float32(0)
	if len(successful) > 0 {
		first := successful[0]
		minLatency = float32(results[first].LatencyMs)
		maxLatency = minLatency
		minCost = results[first].EstimatedCostUSD
		maxCost = minCost
		for _, idx := range successful[1:] {
			lat := float32(results[idx].LatencyMs)
			if lat < minLatency {
				minLatency = lat
			}
			if lat > maxLatency {
				maxLatency = lat
			}
			cost := results[idx].EstimatedCostUSD
			if cost < minCost {
				minCost = cost
			}
			if cost > maxCost {
				maxCost = cost
			}
		}
	}

	for _, idx := range successful {
		r := &results[idx]
		quality := evaluation.QualityScore(r.ReplyPreview, body.RubricKeywords)
		safety := evaluation.SafetyScore(&r.Guardrail)
		latency := evaluation.NormalizedScore(float32(r.LatencyMs), minLatency, maxLatency, true)
		cost := evaluation.NormalizedScore(r.EstimatedCostUSD, minCost, maxCost, true)
		r.Score = models.ProviderBenchmarkScore{
			Quality: quality,
			Latency: latency,
			Cost:    cost,
			Safety:  safety,
			Overall: evaluation.OverallBenchmarkScore(quality, safety, latency, cost),
		}
	}

	sort.SliceStable(results, func(i, j int) bool {
		return results[i].Score.Overall > results[j].Score.Overall
	})

	var recommended *uuid.UUID
	for _, r := range results {
		if r.Error == nil {
			id := r.ProviderID
			recommended = &id
			break
		}
	}

	writeJSON(w, http.StatusOK, models.ProviderBenchmarkResponse{
		BenchmarkGroupID:        benchmarkGroupID,
		UseCase:                 body.UseCase,
		PromptExcerpt:           summarizeTitle(body.Prompt),
		RequiredModalities:      required,
		RequestedPrivateNetwork: body.RequirePrivateNetwork,
		RecommendedProviderID:   recommended,
		Results:                 results,
		CreatedAt:               time.Now().UTC(),
	})
}

// EvaluateGuardrails handles `POST /api/v1/guardrails/evaluate`. Pure
// logic — runs llm.EvaluateText then derives the recommendations
// list mirroring the Rust BTreeSet (sorted set) ordering.
func (h *Handlers) EvaluateGuardrails(w http.ResponseWriter, r *http.Request) {
	var body models.EvaluateGuardrailsRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if strings.TrimSpace(body.Content) == "" {
		writeError(w, http.StatusBadRequest, "guardrail evaluation requires content")
		return
	}

	verdict := llm.EvaluateText(body.Content)

	recSet := map[string]struct{}{}
	if verdict.Blocked {
		recSet["Remove prompt-injection or toxic content before retrying."] = struct{}{}
	}
	for _, flag := range verdict.Flags {
		if strings.HasPrefix(flag.Kind, "pii_") {
			recSet["Redact PII before routing prompts to external LLM providers."] = struct{}{}
		}
	}
	if len(recSet) == 0 {
		recSet["No blocking issues detected; response is safe to continue."] = struct{}{}
	}
	recommendations := make([]string, 0, len(recSet))
	for k := range recSet {
		recommendations = append(recommendations, k)
	}
	sort.Strings(recommendations)

	writeJSON(w, http.StatusOK, models.EvaluateGuardrailsResponse{
		Verdict:         verdict,
		RiskScore:       evaluation.RiskScore(&verdict),
		Recommendations: recommendations,
	})
}
